cli/flox-rust-sdk/src/models/manifest.rs

```diff
@@ -3,11 +3,207 @@ use std::process::Command;
 use std::str::FromStr;
 
 use log::debug;
-use serde::Deserialize;
+use serde::de::Error;
+use serde::{Deserialize, Serialize};
 use toml_edit::{self, DocumentMut, Formatted, InlineTable, Item, Table, Value};
 
+use crate::data::{System, Version};
 use crate::models::pkgdb::PKGDB_BIN;
 
+/// A wrapper around a [`toml_edit::DocumentMut`]
+/// that allows modifications of the raw manifest document,
+/// while preserving comments and user formatting.
+#[derive(Debug)]
+struct RawManifest(toml_edit::DocumentMut);
+impl RawManifest {
+    /// Get the version of the manifest, if it's present or default to v1.
+    fn get_version(&self) -> i64 {
+        self.0
+            .get("version")
+            .and_then(Item::as_integer)
+            .unwrap_or(1)
+    }
+
+    /// Serde's error messages for _untagged_ enums are rather bad
+    /// and don't appear to become better any time soon:
+    /// - <https://github.com/serde-rs/serde/pull/1544>
+    /// - <https://github.com/serde-rs/serde/pull/2376>
+    ///
+    /// This function aims to provide the intermediate version matching on
+    /// the `version` field, and then deserializes the correct version
+    /// of the Manifest explicitly.
+    ///
+    /// <https://github.com/serde-rs/serde/pull/2525> will allow the use of integers
+    /// (i.e. versions) as enum tags, which will allow us to use `#[serde(tag = "version")]`
+    /// and avoid the [Version] field entirely, where the version field is not optional.
+    ///
+    /// Discussion: using a string field as the version tag `version: "1"` vs `version: 1`
+    /// could work today, but is still limited by the lack of an optional tag.
+    fn to_typed(&self) -> Result<TypedManifest, toml_edit::de::Error> {
+        match self.get_version() {
+            2 => Ok(TypedManifest::Catalog(toml_edit::de::from_document(
+                self.0.clone(),
+            )?)),
+            1 => Ok(TypedManifest::Pkgdb(toml_edit::de::from_document(
+                self.0.clone(),
+            )?)),
+            v => {
+                let msg = format!("unsupported manifest version: {v}");
+                Err(toml_edit::de::Error::custom(msg))
+            },
+        }
+    }
```

Comment:
personally, I'd like to add the `insert_*/remove_*/..` functions as methods of this struct, or at least refactor them to take a reference to this struct rather than a mix of documents and content strings.
Since that's slightly larger, i held off it for this first commit that is merely adding the types.